vnext/Microsoft.ReactNative/Utils/BatchingEventEmitter.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "BatchingEventEmitter.h"
+#include "DynamicWriter.h"
+#include "JSValueWriter.h"
+
+namespace winrt::Microsoft::ReactNative {
+
+BatchingEventEmitter::BatchingEventEmitter(Mso::CntPtr<const Mso::React::IReactContext> &&context) noexcept
+    : m_context(std::move(context)) {
+  m_uiDispatcher = m_context->Properties().Get(ReactDispatcherHelper::UIDispatcherProperty()).as<IReactDispatcher>();
+}
+
+void BatchingEventEmitter::EmitJSEvent(int64_t tag, winrt::hstring &&eventName, JSValue &&eventObject) noexcept {
+  return EmitJSEvent(L"receiveEvent", tag, std::move(eventName), std::move(eventObject));
+}
+
+void BatchingEventEmitter::EmitJSEvent(
+    winrt::hstring &&emitterMethod,
+    int64_t tag,
+    winrt::hstring &&eventName,
+    JSValue &&eventObject) noexcept {
+  VerifyElseCrash(m_uiDispatcher.HasThreadAccess());
+
+  implementation::BatchedEvent newEvent{std::move(emitterMethod), tag, std::move(eventName), std::move(eventObject)};
+
+  size_t queueSizeAfterInsert;
```

Comment:
nit, I'd make this a `bool` named `queueWasEmpty` or something like that, since you don't really care about the size, just that it was previously empty.